Introduce primitive threading in unified scheduler

ryoqun opened the pull request:
(no need to start to review this at weekend).

#### Problem

There's no threading support in the unified scheduler, which will exhibit an extremely minimized overhead for unbatched processing.

#### Summary of Changes

Add minimum threading model and wire it, which still lacks suspending (and resuming) threads and aborting them on transaction error in dead blocks. Also, current impl is still full of `.unwrap()`s.

also, note that the actual scheduling code will come in later prs. after that, the code in this pr is ready for multi threaded with proper scheduling code. yet, it's still single threaded at the time of this pr. all this breakdowns are done for bite-sized pr reviewing.

Also, patching crossbeam is planned but yet not included in this pr as well.

#### Context

extracted from: https://github.com/solana-labs/solana/pull/33070

ryoqun commented on unified-scheduler-pool/src/lib.rs:
```diff
@@ -31,6 +35,7 @@ use {
             atomic::{AtomicU64, Ordering::Relaxed},
             Arc, Mutex, Weak,
         },
+        thread::{self, JoinHandle},
```
hi, thread. welcome to the unified scheduler land. lol

ryoqun commented on unified-scheduler-pool/src/lib.rs:
```diff
@@ -194,34 +199,422 @@ impl TaskHandler for DefaultTaskHandler {
+// minimum at the cost of a single heap allocation per switching for the sake of Box-ing the Self
+// type to avoid infinite mem::size_of() due to the recursive type structure. Needless to say, such
+// an allocation can be amortized to be negligible.
+enum ChainedChannel<P1, P2> {
```
originally, i used condvar, but later i noticed this trick and never looked back. :) hope i explained it well in the above comment...

apfitzge replied:
Appreciate the detailed comment here; I think the chained channel idea makes sense. My only real gripe is the visibility to internals within this code, the chained channel is used for synchronization but could easily be screwed up by a change in the code of the handler thread.

It seems possible we can isolate this type and offer an interface for recving that will internally mutate the receiving channel; which will guarantee each receiver can only receive it once (which is currently depending on the handler to overwrite its' receiver).

(pseudo-code with removed generics)
```rust
struct ChainedChannelReceiver<...> {
    receiver: Receiver<...>,
    context: Context,
}

impl ChainedChannelReceiver {
    /// Blocks to receive the next Payload.
    /// If a context-switch is recveived, the context and channel are mutated,
    /// and the next payload message will be blocked on.
    pub fn recv(&mut self) -> Result<Payload> {
        loop {
            match self.receiver.recv()? {
                Payload(p) => return Ok(p),
                PayloadAndChannel((p, new_channel)) => {
                    self.receiver = new_channel;
                    self.context = p;
                }
            }
        }
    }
}

```

apfitzge replied:
Not sure my suggestion works if the handler threads will receive from other channels that would actually require a `select!` though.

ryoqun replied:
nice suggestion :) 9b5e3383d8e7ed251e5ecba33916ab96c35fb6a5

> Not sure my suggestion works if the handler threads will receive from other channels that would actually require a select! though.

this is good point as well. i did my best in the commit.

ryoqun commented on unified-scheduler-pool/src/lib.rs:
```diff
@@ -577,8 +966,10 @@ mod tests {
 
     #[derive(Debug)]
     struct AsyncScheduler<const TRIGGER_RACE_CONDITION: bool>(
-        PooledScheduler<DefaultTaskHandler>,
```
now stop using the impl by wrapping as the impl is getting fleshed out, meaning it's harder to use for a race condition test

ryoqun commented on unified-scheduler-pool/src/lib.rs:
```diff
@@ -194,34 +199,422 @@ impl TaskHandler for DefaultTaskHandler {
 #[derive(Debug)]
 pub struct PooledSchedulerInner<S: SpawnableScheduler<TH>, TH: TaskHandler> {
-    id: SchedulerId,
+    thread_manager: ThreadManager<S, TH>,
```
note that `ThreadManager` is planned to be wrapped in `Arc<Mutex<_>>` soon. so, i didn't flatten it into PooledSchedulerInner.

ryoqun replied:
well well well, I ended up not doing so with different impl.

ryoqun commented on unified-scheduler-pool/src/lib.rs:
```diff
@@ -295,7 +677,7 @@ impl<TH: TaskHandler> InstalledScheduler for PooledScheduler<TH> {
     }
 
     fn pause_for_recent_blockhash(&mut self) {
-        // not surprisingly, there's nothing to do for this min impl!
```
:byebye:

ryoqun commented on unified-scheduler-pool/src/lib.rs:
```diff
@@ -194,34 +199,422 @@ impl TaskHandler for DefaultTaskHandler {
+            move || loop {
+                let mut is_finished = false;
+                while !is_finished {
+                    select! {
```
`select!`'s directive order is logically off intentionally for upcoming use of `select_biased!`..

ryoqun commented on unified-scheduler-pool/src/lib.rs:
```diff
@@ -194,34 +199,422 @@ impl TaskHandler for DefaultTaskHandler {
+
+        let mut result_with_timings = self.session_result_with_timings.take();
+
+        // High-level flow of new tasks:
+        // 1. the replay stage thread send a new task.
+        // 2. the scheduler thread accepts the task.
+        // 3. the scheduler thread dispatches the task after proper locking.
+        // 4. the handler thread processes the dispatched task.
+        // 5. the handler thread reply back to the scheduler thread as an executed task.
+        // 6. the scheduler thread post-processes the executed task.
+        // 7. the scheduler thread send the executed task to the accumulator thread.
+        // 8. the accumulator thread examines the executed task's result and accumulate its timing,
+        //    finally dropping the transaction inside the executed task.
```
across the new code, i used the term task instead of transaction for envisioning bundles as well along with transactions under the generic struct of Task.....

ryoqun commented on unified-scheduler-pool/src/lib.rs:
```diff
@@ -194,34 +199,422 @@ impl TaskHandler for DefaultTaskHandler {
+        );
+    }
+
+    fn start_threads(&mut self, context: &SchedulingContext) {
```
closures in this method is pretty full of unwraps on crossbeam channels. most of them will be addressed in later prs.

that's said, these unwraps should be acceptable even for production. That's because once scheduler is spawn, it never stops the backing threads.

ryoqun commented on unified-scheduler-pool/src/lib.rs:
```diff
@@ -194,34 +199,422 @@ impl TaskHandler for DefaultTaskHandler {
+    session_result_sender: Sender<Option<ResultWithTimings>>,
+    session_result_receiver: Receiver<Option<ResultWithTimings>>,
+    session_result_with_timings: Option<ResultWithTimings>,
+    scheduler_thread: Option<JoinHandle<()>>,
+    handler_threads: Vec<JoinHandle<()>>,
+    accumulator_thread: Option<JoinHandle<()>>,
```
note that these join handles aren't used properly yet in this pr.

apfitzge commented on unified-scheduler-pool/src/lib.rs:
```diff
@@ -194,34 +199,422 @@ impl TaskHandler for DefaultTaskHandler {
+            }
+        };
+
+        // This thread is needed because .accumualte() and drop::<SanitizedTransaction>() both are
```
what is heavy in `SanitizedTransaction::drop`, it should just be `free`-ing some heap allocations, right?

ryoqun replied:
yep, just `free`-ing some heap allocations is _relatively_ heavy for the unified scheduler. Also, `SanitizedTransaction` is rather complex struct.

I observed some visible cpu usage just from `SanitizedTransaction::drop` when casual profiling.

apfitzge commented on unified-scheduler-pool/src/lib.rs:
```diff
@@ -194,34 +199,422 @@ impl TaskHandler for DefaultTaskHandler {
+            let finished_task_sender = finished_task_sender.clone();
+
+            move || loop {
+                let (task, sender) = select! {
```
there's no selection going on here, since there's only one channel receiving?

ryoqun replied:
oh, nice question. `select!` here is intentional because the handler will soon use prioritized queue, which will be composed of two channels: 1. low priority tasks (not conflicting with others), 2. high priority tasks (conflicting with others).

so, I'll leave this rather odd select! as-is in this pr.

ryoqun replied:
> the handler will soon use prioritized queue, ....

_soon_ :trollface: https://github.com/anza-xyz/agave/pull/627#discussion_r1554626991

ryoqun commented on unified-scheduler-pool/src/lib.rs:
```diff
@@ -563,7 +946,13 @@ mod tests {
+        // However, currently threaded impl can't properly abort in this situtation..
+        // so, 1 should be observed, intead of 0.
+        // Also note that bank.transaction_count() is generally racy because blockstore_processor
+        // and unified_scheduler both tend to process non-conflicting batches in parallel as normal
+        // operation.
```
wording: `as part of normal functioning`

ryoqun commented on unified-scheduler-pool/src/lib.rs:
```diff
@@ -194,34 +199,422 @@ impl TaskHandler for DefaultTaskHandler {
     }
 }
 
+pub struct ExecutedTask {
```
nit: remove `pub`

ryoqun replied:
done: a2eb73e1346a0731651d6770acd5312bb3871781

ryoqun commented on unified-scheduler-pool/src/lib.rs:
```diff
@@ -194,34 +199,422 @@ impl TaskHandler for DefaultTaskHandler {
+        // 4. the handler thread processes the dispatched task.
+        // 5. the handler thread reply back to the scheduler thread as an executed task.
+        // 6. the scheduler thread post-processes the executed task.
+        // 7. the scheduler thread send the executed task to the accumulator thread.
+        // 8. the accumulator thread examines the executed task's result and accumulate its timing,
```
yeah, it is expensive. any offload-able work should be put elsewhere from the scheduling thread: beae8610dd71456adcebdebad9ac512d3388911a

apfitzge commented on unified-scheduler-pool/src/lib.rs:
```diff
@@ -194,34 +199,535 @@ impl TaskHandler for DefaultTaskHandler {
+            // Moreover, the scheduler is designed to handle tasks without batching at all in the
+            // pursuit of saturating all of the handler threads with maximally-fine-grained
+            // concurrency density for throughput as the second design goal. This design goal
+            // relies on the assumption that there's no considerable penalty arising from the
+            // unbatched manner of processing.
```
To be clear, I'm not strictly opposed to eliminating batching in replay.
It's probable that the gained parallelism from not doing batches can, in many cases, make it overall more efficient.

But I think we shouldn't state the assumption there's no significant loss, when the data suggests otherwise.

ryoqun replied:
> 20us/tx => 1280us/(64tx). So there's pretty significant loss of efficiency by eliminating batching entirely.

well, i think i managed to eliminate the overhead: https://github.com/apfitzge/solana/pull/3/files

tested on old intel cpu

```
running 8 tests
test bench_execute_batch_01_txes_per_batch          ... bench:     199,425 ns/iter (+/- 21,345)
test bench_execute_batch_04_txes_per_batch          ... bench:     203,539 ns/iter (+/- 31,947)
test bench_execute_batch_16_txes_per_batch          ... bench:     211,507 ns/iter (+/- 72,429)
test bench_execute_batch_64_txes_per_batch          ... bench:     180,725 ns/iter (+/- 36,828)
test bench_execute_batch_flattend_01_txes_per_batch ... bench:     182,158 ns/iter (+/- 25,963)
test bench_execute_batch_flattend_04_txes_per_batch ... bench:     288,362 ns/iter (+/- 13,242)
test bench_execute_batch_flattend_16_txes_per_batch ... bench:     181,084 ns/iter (+/- 17,782)
test bench_execute_batch_flattend_64_txes_per_batch ... bench:     185,192 ns/iter (+/- 26,854)
```

tested on recent amd cpu:

```
running 8 tests
test bench_execute_batch_01_txes_per_batch          ... bench:     210,430 ns/iter (+/- 4,364)
test bench_execute_batch_04_txes_per_batch          ... bench:     177,109 ns/iter (+/- 5,016)
test bench_execute_batch_16_txes_per_batch          ... bench:     188,921 ns/iter (+/- 7,451)
test bench_execute_batch_64_txes_per_batch          ... bench:     180,847 ns/iter (+/- 5,388)
test bench_execute_batch_flattend_01_txes_per_batch ... bench:     183,689 ns/iter (+/- 4,730)
test bench_execute_batch_flattend_04_txes_per_batch ... bench:     183,313 ns/iter (+/- 7,748)
test bench_execute_batch_flattend_16_txes_per_batch ... bench:     176,748 ns/iter (+/- 1,863)
test bench_execute_batch_flattend_64_txes_per_batch ... bench:     177,591 ns/iter (+/- 4,527)

```

ryoqun replied:
in that dirtiest branch, i disabled most of non-essential code, which is causing those overheads. i think all of them are fixable to have no overhead with moderate amount efforts.

indeed, unbatched code path incurs raw increase of instruction count. however, these are basically can be offset for better data/code locality. i can't write in detail (it's midnight here..). but i confirmed this with amd/intel with `perf stat`.

so, this is a glimpse of what I'm arguing against batching for years. :)

ryoqun commented on unified-scheduler-pool/src/lib.rs:
```diff
@@ -194,34 +200,529 @@ impl TaskHandler for DefaultTaskHandler {
+    }
+
+    // P doesn't need to be `: Clone`, yet rustc derive can't handle it.
+    // see https://github.com/rust-lang/rust/issues/26925
```
ref: https://github.com/rust-lang/rust/issues/26925

ryoqun commented on unified-scheduler-pool/src/lib.rs:
```diff
@@ -194,34 +199,535 @@ impl TaskHandler for DefaultTaskHandler {
+
+            // Now, this is the main loop for the scheduler thread, which is a special beast.
+            //
+            // That's because it's the most notable bottleneck of throughput. Unified scheduler's
+            // overall throughput is largely dependant on its ultra-low latency characteristic,
+            // which is the most important design goal of the scheduler in order to reduce the
+            // transaction confirmation latency for end users.
```
> It might be the bottleneck of just scheduling, but the main bottleneck of replay will still be the processing of transactions (load, execute, commit), right?

Currently, yes. given ~100 handler threads, scheduler thread can be the bottleneck. and I'm developing this scheduler with that future in mind.

> Correct me if I'm wrong, but it seems the sensitiviy is there because between the time of ExecutedTask(Tx0) and Task(Tx1), the handler thread is idle.
> Even with the additional thread which does the accumulation and dropping, there is still going to be some amount of idle time here.

yeah. these are correct.

> Just as a note, I also ran into this in the block-production scheduler in core, but opted to allow the queueing of txs for specific handlers. Even trying to keep just 2 txs always enqueued gives significant leeway in the latency, since as long as the scheduler can queue another in the time it takes to process a transaction there is nearly no idling time.

hmm, this seems odd. busy looping at the other hand of mpmc channels usually gives round-trip less-than-500ns latency, which i'm planning to employ for unified scheduler. i think those threads were yielding to the kernel (ie parking)?

as you pointed out, i can't eliminate the idle time. but then, 1us (500ns for channel + 500ns for scheduling) per a single conflicting tx is acceptable latency.

actually, i wanted to show this as some bench, but i'm running out of time.. i spend too much on the other concern... lol https://github.com/solana-labs/solana/pull/34676#discussion_r1450581025

apfitzge replied:
> i think those threads were yielding to the kernel (ie parking)?

possible.

> as you pointed out, i can't eliminate the idle time. but then, 1us (500ns for channel + 500ns for scheduling) per a single conflicting tx is acceptable latency.

what's the added latency from dropping the txs? or in other words, what would be the affect on overall throughput if we did not have the separate "drop" thread? My main concern here was just that we're spawning this new thread as part of the scheduler, and if that's necessary we should be able to back that decision up with some numbers on how much it helps (or how much we are hurt by **not** doing that).

> as you pointed out, i can't eliminate the idle time.

yeah I think without allowing conflicts to actually queue up, this cannot be resolved entirely, just mitigated.

ryoqun replied:
thanks for detailing your concern.

i've finally wrote benches: 6120684e5933a908018cde4d4ae930ea5448de88 (won't push to this pr)....:

```
test bench_tx_throughput_drop_in_scheduler ... bench:   8,556,627 ns/iter (+/- 3,168,724)
test bench_tx_throughput_drop_in_accumulator ... bench:   7,334,701 ns/iter (+/- 2,600,740)
test bench_tx_throughput_drop_in_scheduler_conflicting ... bench:  11,403,197 ns/iter (+/- 3,062,017)
test bench_tx_throughput_drop_in_accumulator_conflicting ... bench:  10,794,434 ns/iter (+/- 6,808,692)
```

so, processing 10,000 txes takes always longer when we drop txes in the scheduler thread, regardless conflicting or not.

note that you have to ensure to rebuild when repoto these results.. cargo usually skips rebuild with feature unification...

apfitzge replied:
The raw numbers looks like dropping in scheduler takes longer, which isn't surprising since its doing more work; but they are well within the statistical noise. 

`bench_tx_throughput_drop_in_scheduler_conflicting` raw shows 609us higher than `bench_tx_throughput_drop_in_accumulator_conflicting`, but the error bounds are 3ms and 6ms. 

This seems to imply that there's *not* a statistically significant effect from having the separate accumulator thread. Am I missing something in these results?

ryoqun replied:
I appreciate the feedback. Seems casual benchmark wasn't enough.

I reverted the accumulator thread (https://github.com/solana-labs/solana/pull/34676/commits/2f592f5a84dba4fa61f92e1c23d1923c7bfea238) for now to land this pr.

Now that I've done somewhat extended synthesized bench-marking on this pr. That said, I still think the accumulator thread performs better in the end.

However, because this pr doesn't contain any actual scheduling impl, i concluded that presenting some convincing numbers is rather difficult. ;) So, I'll try to ship the scheduling impl first, instead of finishing up the scheduler pool (was original plan). Even later, I'll try to land the accumulator thread along with other missing things in the scheduler pool with more proper justification / more realistic benchmark numbers. 

In other words, landing the accumulator thread wasn't essential for now. So, I was too early to ship it.

> Am I missing something in these results?

yea. blame me for not explaining enough. didn't expect the accumulator thread to be so hot...

1. the accumulator thread will contain other offloaded tasks.

Namely, it's used for dumping all detailed timings for individual transactions to influxdb at background fashion (the code is in the base pr). Also, I'm planning to move rpc related blockstore writes here, as this isn't essential for bank hash. These aren't happening in this pr, so the accumulator thread itself was just ahead of it.

2. dropping transactions can be really heavy

in the intention of a quick sanity check bench, i wasn't bothered to do so. but transactions can be really large with many instructions with each having instruction data blobs, pubkeys...

when blocks contain these transactions (it does usually), it clearly throttles the scheduler thread when dropped inside that.

3. medians inside overlapping deviation range don't necessarily mean statistical noise.

firstly, these multi-threaded benches tend to contain high variance in runtime duration. so, i did little attention to the deviations. I just shared the numbers after some casual multiple runs, thinking these numbers are relatively stable enough.


maybe, seeing is believing. :) here's comparable criterion bench results, which i wrote with different parameters/settings:

![image](https://github.com/solana-labs/solana/assets/117807/1df5d69f-38b8-439a-9b1f-748d24024bc9)

![image](https://github.com/solana-labs/solana/assets/117807/175145f1-287e-49ab-b6ae-ce2e76ad7ed8)

note that I think these periodic outliers are due to cpu freq scaling. Sans these outlifers, I think the primary convex of pdf is shifted in favor of dropping in accumlator, clearly showing statistically significant improvement. Also, this is visibly confirmed by plotted individual samples.

> I think we are not really seeing eye-to-eye on this additional thread, and might just be we have different perspectives.

i think we have a constructive discussion. thanks for patience and expressing your perspective.

> Feel free to pull another core contributor in for their opinion on this aspect of the PR.

thanks for suggestion. yeah, we could seek other's opinions, should we be stuck in other topics.

> I don't recall seeing anything else that would block this PR for me, but will take another look once the thread-issue is settled.

that's good to hear. :)

ryoqun commented:
@apfitzge thanks for review! I've replied to you: https://github.com/solana-labs/solana/pull/34676#discussion_r1457609123

also, i pushed new commits: https://github.com/solana-labs/solana/pull/34676/commits/cc5a1f071fd95bdf624308936b509856ce8aee4b and rebase was needed to pick 0e8f2de1da2e38451dfd4cc45aa51d10dfe428f1

apfitzge commented on unified-scheduler-pool/src/lib.rs:
```diff
@@ -194,34 +200,462 @@ impl TaskHandler for DefaultTaskHandler {
+}
+
+// A very tiny generic message type to signal about opening and closing of subchannels, which are
+// logically segmented series of Payloads (P1) over a signle continuous time-span, potentially
```
```suggestion
// logically segmented series of Payloads (P1) over a single continuous time-span, potentially
```

ryoqun replied:
oops, nice!: 07de5b4736eb811a7943aa97a43fd4ae4fa892a2

apfitzge commented on unified-scheduler-pool/src/lib.rs:
```diff
@@ -593,25 +1025,25 @@ mod tests {
     {
         fn id(&self) -> SchedulerId {
-            self.0.id()
+            todo!();
```
probably better to use `unimplemented!()`, I don't believe you are intending to do this, right?

ryoqun replied:
you're right: afd580e8f86634447deaa86282e24bd257fbdee5

apfitzge commented on unified-scheduler-pool/src/lib.rs:
```diff
@@ -194,34 +200,462 @@ impl TaskHandler for DefaultTaskHandler {
+            // cycles out of the scheduler thread. Thus, any kinds of unessential overhead sources
+            // like syscalls, VDSO, and even memory (de)allocation should be avoided at all costs
+            // by design or by means of offloading at the last resort.
+            move || loop {
```
There's no exit from this loop aside from crashing as far as I can tell.
If replay stage is shutting down, channels get disconnected on the send side and we'd hit Errs in the unwrap - instead of the scheduler loop cleaning shutting down.

Feel free to address in a follow-up PR, as I think you mentioned you were planning already.

ryoqun replied:
your analysis is right.

> Feel free to address in a follow-up PR, as I think you mentioned you were planning already.

yeah. I'll do. but sadly, solana-validator/solana-ledger-tool's process termination is broken for now to begin with. so, the cascading panics won't happen actually. that's why ci is green, maybe.

namely, there's some cyclic link which prevents bank_forks from being dropped. i have a fix in #33070. note that this bug quite low-priority, imo. there's more deeper problems:  solana-validator's exit subcommand and solana-ledger-tool themselvs aren't trying to do sane shutdown cleaning up.

so, I won't address before i ship the scheduler code at least.

apfitzge reviewed (approved):
Just a couple of minor comments, feel free to just do them as follow-up PRs.
Appreciate all the back-and-forth on the PR!

ryoqun commented:
> Just a couple of minor comments, feel free to just do them as follow-up PRs. Appreciate all the back-and-forth on the PR!

thanks for timely review as always. according to the discord talk, this pr is blocked (https://discord.com/channels/428295358100013066/910937142182682656/1197716176764153966) and i just addressed nits and re-requested another code review.

ryoqun commented:
> Sorry my slowness made us miss the cutoff!

well, i don't think so. as i said above, your reviews were very timely. missing the cutoff was mainly due to my hacking and benching, resulting in extended back-and-forth..

seems v1.18 branching is imminent (https://discord.com/channels/428295358100013066/910937142182682656/1199023180598218803), so i just humbly request another approval other then a normal comment. :)

apfitzge reviewed (approved):
lgtm

ryoqun commented on unified-scheduler-pool/src/lib.rs:
```diff
@@ -194,34 +200,462 @@ impl TaskHandler for DefaultTaskHandler {
+            let mut active_task_count: usize = 0;
+
+            // Now, this is the main loop for the scheduler thread, which is a special beast.
+            //
+            // That's because it could be the most notable bottleneck of throughput in the future
+            // when there are ~100 handler threads. Unified scheduler's overall throughput is
+            // largely dependant on its ultra-low latency characteristic, which is the most
+            // important design goal of the scheduler in order to reduce the transaction
+            // confirmation latency for end users.
+            //
+            // Firstly, the scheduler thread must handle incoming messages from thread(s) owned by
+            // the replay stage or the banking stage. It also must handle incoming messages from
+            // the multi-threaded handlers. This heavily-multi-threaded whole processing load must
+            // be coped just with the single-threaded scheduler, to attain ideal cpu cache
+            // friendliness and main memory bandwidth saturation with its shared-nothing
+            // single-threaded account locking implementation. In other words, the per-task
+            // processing efficiency of the main loop codifies the upper bound of horizontal
+            // scalability of the unified scheduler.
+            //
+            // Moreover, the scheduler is designed to handle tasks without batching at all in the
+            // pursuit of saturating all of the handler threads with maximally-fine-grained
+            // concurrency density for throughput as the second design goal. This design goal
+            // relies on the assumption that there's no considerable penalty arising from the
+            // unbatched manner of processing.
+            //
+            // Note that this assumption isn't true as of writing. The current code path
+            // underneath execute_batch() isn't optimized for unified scheduler's load pattern (ie.
+            // batches just with a single transaction) at all. This will be addressed in the
+            // future.
+            //
+            // These two key elements of the design philosophy lead to the rather unforgiving
+            // implementation burden: Degraded performance would acutely manifest from an even tiny
+            // amount of individual cpu-bound processing delay in the scheduler thread, like when
+            // dispatching the next conflicting task after receiving the previous finished one from
+            // the handler.
+            //
+            // Thus, it's fatal for unified scheduler's advertised superiority to squeeze every cpu
+            // cycles out of the scheduler thread. Thus, any kinds of unessential overhead sources
+            // like syscalls, VDSO, and even memory (de)allocation should be avoided at all costs
+            // by design or by means of offloading at the last resort.
```
here

ryoqun commented on unified-scheduler-pool/src/lib.rs:
```diff
@@ -237,53 +671,42 @@ pub trait SpawnableScheduler<TH: TaskHandler>: InstalledScheduler {
-            result_with_timings: Mutex::new((Ok(()), ExecuteTimings::default())),
-        }
+    fn from_inner(mut inner: Self::Inner, context: SchedulingContext) -> Self {
+        inner.thread_manager.start_session(&context);
```
this

ryoqun commented on unified-scheduler-pool/src/lib.rs:
```diff
@@ -194,34 +200,462 @@ impl TaskHandler for DefaultTaskHandler {
+        // single-threaded still.
+        let handler_count = 1;
+
         Self::from_inner(
```
this
